optax/_src/loss.py

```diff
@@ -252,3 +252,99 @@ def log_cosh(
   errors = (predictions - targets) if (targets is not None) else predictions
   # log(cosh(x)) = log((exp(x) + exp(-x))/2) = log(exp(x) + exp(-x)) - log(2)
   return jnp.logaddexp(errors, -errors) - jnp.log(2.0).astype(errors.dtype)
+
+
+def ctc_loss(logits: chex.Array,
+             logitpaddings: chex.Array,
+             labels: chex.Array,
+             labelpaddings: chex.Array,
+             blank_id: int = 0,
+             logepsilon: float = -1e5) -> chex.Array:
+  """Computes CTC loss.
+
+  This function performs forward computation over an FSA with `N * 2` states
+  where `N` is the max number of labels. The states are split into two groups:
+  Phi states and emission states. a phi-state accepts repetition of
+  phi (blank)-symbols and transits to emission state when the correct label is
+  observed. An emission state accepts repetition of the label and transits to
+  the next phi states at any time (so called epsilon-transition).
+  Below, `B` denotes the batch size, `T` denotes the time steps in `logits`,
+  and `N` denotes the time steps in `labels`.
+
+  Args:
+    logits: (B, T, K)-array containing log-probabilities of each class.
+    logitpaddings: (B, T)-array. Padding indicators for `logits`.
+    labels: (B, N)-array containing reference integer labels.
+    labelpaddings: (B, N)-array. Padding indicators for `labels`. Currently,
+      `labels` must be right-padded, i.e. each row of `labelpaddings` must be
+      repetition of zeroes, followed by repetition of ones.
+    blank_id: Id for blank token.
+
+  Returns:
+    (B,)-array containing loss values for each sequence in the batch.
+  """
+  batchsize, unused_maxinputlen, num_classes = logits.shape
+  batchsize_of_labels, maxlabellen = labels.shape
+  chex.assert_equal(batchsize, batchsize_of_labels)
+
+  logprobs = jax.nn.log_softmax(logits)
+  labellens = maxlabellen - jnp.sum(labelpaddings, axis=1).astype(jnp.int32)
+
+  # repeat[b, n] == 1.0 when label[b, n] == label[b, n+1].
+  repeat = (labels[:, :-1] == labels[:, 1:]).astype(jnp.float32)
+  repeat = jnp.pad(repeat, ((0, 0), (0, 1)))
+
+  logprobs_phi = logprobs[:, :, blank_id:blank_id + 1]  # [B, T, 1]
+  logprobs_phi = jnp.transpose(logprobs_phi, (1, 0, 2))  # [T, B, 1]
+
+  one_hot = jax.nn.one_hot(labels, num_classes=num_classes)  # [B, N, K]
+  logprobs_emit = jnp.einsum('btk,bnk->btn', logprobs, one_hot)
+  logprobs_emit = jnp.transpose(logprobs_emit, (1, 0, 2))  # [T, B, N]
+
+  logalpha_phi_init = jnp.ones(
+      (batchsize, maxlabellen + 1)) * logepsilon  # [B, N]
+  logalpha_phi_init = logalpha_phi_init.at[:, 0].set(0.0)
+  logalpha_emit_init = jnp.ones(
+      (batchsize, maxlabellen)) * logepsilon  # [B, N]
+
+  def loop_body(prev, x):
+    prev_phi, prev_emit = prev
+    # emit-to-phi epsilon transition, except if the next label is repetition
+    prev_phi_orig = prev_phi
+    prev_phi = prev_phi.at[:, 1:].set(
```

Comment:
Out of curiosity, did you run any benchmarks to measure the efficiency of this operation? I heard that .set() is very slow, and, since you're updating a contiguous subtensor, a simple concatenation might work a way faster.

E.g. you could have a helper function

```
def _update_1up(phi, update):
   return jnp.concatenate([phi[:, :1], update], axis=-1)
```

Reply:
Thanks for the suggestion, done.
In my benchmark script, it improved the performance by >30%,
specifically from ~27ms to ~20ms in the setting with `logits.shape==(128, 1024, 64)`, and `labels.shape==(128, 256)`.